### DISCUSSION TITLE
Adding MacOS M1 Metal GPU Tutorial

### DISCUSSION THREAD

**Discussion by Killpit** [mrdbourke/pytorch-deep-learning#337] [Ideas]
Hi Dan, 

I'd recommend you to add comprehensive MacBook M1 Metal GPU Tutorial for MacBook M1 Metal GPU users. I looked at the documentation and couldn't find anything neither on Apple or PyTorch documentation.

**Comment by mrdbourke** [comment]
Hi @Killpit,

Thank you for the suggestion! 

But could you be more specific?

What would you like to do on MX Metal chips?

Have you seen this: https://developer.apple.com/metal/pytorch/ 

I also have a video on setting up PyTorch for Mac: https://youtu.be/Zx2MHdRgAIc

**Reply by Killpit** [reply]
No, I set up PyTorch on Mac, but I couldn't find how can I access the MetalGPU on MacOS, whatever I can find is only CUDA. Also I followed both PyTorch documentation and Apple developer documentation you sent below, they taught quite a lot

**Reply by mrdbourke** [reply]
Hi @Killpit, you can access the Apple GPU via the following:

```python
# Setup device-agnostic code 
if torch.cuda.is_available():
    device = "cuda" # NVIDIA GPU
elif torch.backends.mps.is_available():
    device = "mps" # Apple GPU
else:
    device = "cpu" # Defaults to CPU if NVIDIA GPU/Apple GPU aren't available

print(f"Using device: {device}")
```

See more here: https://www.learnpytorch.io/pytorch_cheatsheet/#device-agnostic-code-using-pytorch-on-cpu-gpu-or-mps

**Reply by Killpit** [reply]
Thanks for the help, I'll use this part of code, the documentation was really confusing but I don't really have to worry about it.

**Reply by Killpit** [reply]
@mrdbourke does your GPU tutorial code works with PyCharm for MPS

**Reply by mrdbourke** [reply]
yes, it should work wherever you need to use Apple GPUs (and have the right code/environment installation setup)

**Comment by joma94** [comment]
Following your brilliant class,- everything runs perfect on my Macbook Air M2. (24 Gb ram 2T ssd)
Equal to - or faster than Google Colab with GPU.

Until 04. PyTorch Custom Datasets - part 7.7 Train and Evaluate Model 0.

Colab does each epoch in a couple of seconds,- my M2 takes a little over a minute per epoch.

I haven't figured that one out yet :) Yes, everything says "mps" - but ... no clue :)
First warning about something a bit sideways in "7.3 Try a forward pass on a single image (to test the model)" - it takes 25 secs.

I can reduce the time per epoch by using BATCH_SIZE=512 to around 15 sec's.
Further increase 1024 - 2048 does not help further.
